### PR TITLE
Added "Snap to date with highest coverage" checkbox to search results page.

### PR DESCRIPTION
**Pull request by dwnoble**
Note: Needs unit tests

- Added `Snap to date with highest coverage` checkbox.
- Added i18n support to subject, explore, topic, and NL pages.
- Added `AbortController` to  place_name.tsx component to gracefully handle unmounting before loading is complete (this can happen when toggling the  "Snap to date with highest coverage" checkbox before the chart has finished loading)

Heuristic for fetching highest coverage observations:
Choose the date with the most data coverage from either:
(1) last 5 observation dates
(2) 5 years from the most recent observation date
whichever set is greater

For (1) uses the `/v1/bulk/observation-dates/linked` API to get entity counts by date and facet. 

Checked (new default)
Screenshot 2024-02-29 at 4 30 15 PM

Unchecked
Screenshot 2024-02-29 at 4 30 25 PM

Info hover
Screenshot 2024-02-29 at 4 31 12 PM

### CODE REVIEW

File: server/lib/util.py [hunk 1]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
+
+
+def fetch_highest_coverage(parent_entity: str,
+                           child_type: str,
+                           variables: List[str],
+                           all_facets: bool,
+                           facet_ids: List[str] = None):
+  """
+  Fetches the latest available data with the best coverage for the given a
+  parent entity, child type, variables, and facets
```
Comment (shifucun) [inline server/lib/util.py:515]: To clarify, is the "best coverage" on a per stat var basis?

File: server/lib/util.py [hunk 2]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
+    # Heuristic to fetch the "FETCH_HIGHEST_COVERAGE_FROM_DATES" most recent
+    # observation dates or observation dates going back
+    # "FETCH_HIGHEST_COVERAGE_FROM_YEARS" years, whichever is greater
+    cutoff_year = str(date.today().year - FETCH_HIGHEST_COVERAGE_FROM_YEARS)
+    latest_observation_dates_from_year = [
+        o for o in descending_observation_dates if o['date'] > cutoff_year
+    ]
+    latest_observation_dates = descending_observation_dates[:
+                                                            FETCH_HIGHEST_COVERAGE_FROM_DATES]
+    observation_dates = latest_observation_dates_from_year if len(
```
Comment (shifucun) [inline server/lib/util.py:564]: why comparing the length not the latest date from the two list?

File: server/lib/util.py [hunk 3]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
+  # Choose the date with the most data coverage from either:
+  # (1) last 5 observation dates
+  # (2) 5 years from the most recent observation date
+  # whichever set is greater
+  FETCH_HIGHEST_COVERAGE_FROM_DATES = 5
+  FETCH_HIGHEST_COVERAGE_FROM_YEARS = 5
+  point_responses = []
+  series_dates_response = dc.get_series_dates(parent_entity, child_type,
+                                              variables)
+  facet_ids_set = set(facet_ids or [])
```
Comment (chejennifer) [inline server/lib/util.py:541]: nit: can just set default value as [] in the params?

File: server/lib/util.py [hunk 4]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
+    if len(descending_observation_dates) == 0:
+      continue
+    # Heuristic to fetch the "FETCH_HIGHEST_COVERAGE_FROM_DATES" most recent
+    # observation dates or observation dates going back
+    # "FETCH_HIGHEST_COVERAGE_FROM_YEARS" years, whichever is greater
+    cutoff_year = str(date.today().year - FETCH_HIGHEST_COVERAGE_FROM_YEARS)
+    latest_observation_dates_from_year = [
+        o for o in descending_observation_dates if o['date'] > cutoff_year
+    ]
+    latest_observation_dates = descending_observation_dates[:
```
Comment (chejennifer) [inline server/lib/util.py:562]: nit: lines 562 to 566 could be replaced with: 

```
obs_dates_cutoff = max(len(latest_observation_dates_from_year), FETCH_HIGHEST_COVERAGE_FROM_DATES)
observation_dates = descending_observation_dates[:obs_dates_cutoff]
```

File: server/lib/util.py [hunk 5]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
+      }
+    }
+  }
+  """
+  # Heuristic for fetching "latest data with highest coverage":
+  # Choose the date with the most data coverage from either:
+  # (1) last 5 observation dates
+  # (2) 5 years from the most recent observation date
+  # whichever set is greater
+  FETCH_HIGHEST_COVERAGE_FROM_DATES = 5
```
Comment (chejennifer) [inline server/lib/util.py:536]: nit: these constant names are slightly confusing, maybe just 'MAX_DATES_TO_CHECK' and 'MAX_YEARS_TO_CHECK'?

File: static/css/shared/_ranking_unit.scss [hunk 6]
```diff
@@ -18,11 +18,18 @@
  * Styling for ranking unit component.
  */
 
+.ranking-tile {
+  .chart-footnote {
+    margin: 8px 0 0 0;
+    font-weight: 500;
```
Comment (juliawu) [inline static/css/shared/_ranking_unit.scss:24]: Nit: Could you alphabetize the properties here and in the other scss blocks you edited?

File: server/lib/util.py [hunk 7]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
+            max([
+                entity_count_item
+                for entity_count_item in obs['entityCount']
+                if (all_facets or len(facet_ids_set) == 0 or
+                    entity_count_item['facet'] in facet_ids_set)
+            ],
+                key=lambda item: item['count'])['count']
+    }
+                   for obs in observation_dates]
+    best_coverage = max(date_counts, key=lambda date_count: date_count['count'])
```
Comment (chejennifer) [inline server/lib/util.py:583]: nit: have a helper function for getting the best date, given a list of observations

File: server/lib/util.py [hunk 8]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
+  FETCH_HIGHEST_COVERAGE_FROM_YEARS = 5
+  point_responses = []
+  series_dates_response = dc.get_series_dates(parent_entity, child_type,
+                                              variables)
+  facet_ids_set = set(facet_ids or [])
+  for variable_date_item in series_dates_response['datesByVariable']:
+    # Each variable_date_item contains the observation counts by date
+    variable = variable_date_item['variable']
+    # Get observation dates in descending order, and filter out dates in the
+    # future (to exclude erroneous data)
```
Comment (juliawu) [inline server/lib/util.py:546]: I think in the current implementation "future" means just in the next year, right? Does the logic also exclude future dates within the current year (e.g. Today is March 4, 2024 and the observation date is April 4, 2024)? If not, is that a case we want to handle?

File: server/lib/util.py [hunk 9]
```diff
@@ -500,3 +503,89 @@ def gzip_compress_response(raw_content, is_json):
...
   return response
+
+
+def fetch_highest_coverage(parent_entity: str,
+                           child_type: str,
+                           variables: List[str],
+                           all_facets: bool,
+                           facet_ids: List[str] = None):
+  """
+  Fetches the latest available data with the best coverage for the given a
```
Comment (chejennifer) [inline server/lib/util.py:514]: it looks like each variable could be using a different date. Could you note that here

File: shared/lib/constants.py [hunk 10]
```diff
@@ -459,3 +459,10 @@
 
 # English language code.
 EN_LANG_CODE = 'en'
+
+# date query param value for requesting latest observations from REST v2 API
+DATE_LATEST = 'LATEST'
+
+# date query param value for requesting latest overall available for
```
Comment (chejennifer) [inline shared/lib/constants.py:466]: nit: note that the difference between DATE_LATEST and DATE_HIGHEST_COVERAGE is that one will return the same date for all the data and the other could have different dates?

File: static/js/apps/explore/app.tsx [hunk 11]
```diff
@@ -124,44 +126,50 @@ export function App(props: { isDemo: boolean }): JSX.Element {
...
-          query={query}
-          debugData={debugData}
-          exploreContext={exploreContext}
-          queryResult={queryResult}
-          pageMetadata={pageMetadata}
-          userMessage={userMessage}
-        />
-      )}
-    </Container>
+    <RawIntlProvider value={intl}>
```
Comment (chejennifer) [inline static/js/apps/explore/app.tsx:129]: mention in the PR description that you're also adding i18n support in this PR

**Review by juliawu — COMMENTED** [review]
Awesome work! Thanks for adding the abort controller and i18n plumbing as well.

File: static/js/components/place_name.tsx [hunk 12]
```diff
@@ -33,9 +34,16 @@ export function PlaceName(props: PlaceNameProp): JSX.Element {
   // parent place is USA
   const [name, setName] = useState<string>("");
   useEffect(() => {
+    const controller = new AbortController();
```
Comment (chejennifer) [inline static/js/components/place_name.tsx:37]: also mention in PR description about the abort controller/what the purpose is?

File: static/js/components/subject_page/block.tsx [hunk 13]
```diff
@@ -129,18 +149,70 @@ export function Block(props: BlockPropType): JSX.Element {
...
+    _.flatten(props.columns.map((c) => c.tiles.map((tile) => tile.statVarKey)))
+  );
+  const tileTypes = _.flatten(
+    _.flatten(props.columns.map((c) => c.tiles.map((tile) => tile.type)))
+  );
+  const statVarSpecs = props.statVarProvider.getSpecList(statVarKeys);
+  // Only show the "Snap to date with highest coverage" checkbox if
+  // (1) No date is set (e.g., date is "LATEST")
+  // (2) Chart types are map and/or ranking
+  const showSnapToLatestData =
```
Comment (chejennifer) [inline static/js/components/subject_page/block.tsx:162]: nit: can have a helper for getting whether or not to show snap to latest data

File: static/js/components/subject_page/block.tsx [hunk 14]
```diff
@@ -235,17 +318,21 @@ function renderTiles(
       : props.place;
     const comparisonPlaces = getComparisonPlaces(tile, place);
     const className = classNameList.join(" ");
+    const statVarSpecs = props.statVarProvider.getSpecList(tile.statVarKey, {
```
Comment (chejennifer) [inline static/js/components/subject_page/block.tsx:321]: curious why you moved this here even though most of the switch statements are still making this call?

**Comment by miss-o-soup** [comment]
From a UI presentation perspective, let's left align the text in the tooltip, and thanks for fixing the dates and units wrapping issue: https://screenshot.googleplex.com/8YJW7xxEEitiVPr

File: server/lib/util.py (hunk 1 above)
Comment (dwnoble) [reply server/lib/util.py:515]: Yes, it's per stat var. So a request for multiple stat vars could come back with 2021 for the best coverage for stat var A and 2022 for the best coverage for stat var B

File: server/lib/util.py (hunk 2 above)
Comment (dwnoble) [reply server/lib/util.py:564]: The latest date from the two lists should always be the same (the most recent year). This logic just checks how far to walk backwards in time from the latest date.

File: static/css/shared/_ranking_unit.scss (hunk 6 above)
Comment (dwnoble) [reply static/css/shared/_ranking_unit.scss:24]: done

Comment (dwnoble) [reply static/css/shared/_ranking_unit.scss:24]: Fixed!

File: server/lib/util.py (hunk 8 above)
Comment (dwnoble) [reply server/lib/util.py:546]: Fixed to use today's date

File: server/lib/util.py (hunk 3 above)
Comment (dwnoble) [reply server/lib/util.py:541]: Updated

File: server/lib/util.py (hunk 5 above)
Comment (dwnoble) [reply server/lib/util.py:536]: Good callout-updated

File: server/lib/util.py (hunk 9 above)
Comment (dwnoble) [reply server/lib/util.py:514]: Added a note

File: server/lib/util.py (hunk 7 above)
Comment (dwnoble) [reply server/lib/util.py:583]: Done

File: shared/lib/constants.py (hunk 10 above)
Comment (dwnoble) [reply shared/lib/constants.py:466]: Done

File: static/js/apps/explore/app.tsx (hunk 11 above)
Comment (dwnoble) [reply static/js/apps/explore/app.tsx:129]: done

**Comment by dwnoble** [comment]
> From a UI presentation perspective, let's left align the text in the tooltip, and thanks for fixing the dates and units wrapping issue: https://screenshot.googleplex.com/8YJW7xxEEitiVPr

Done

File: server/lib/util.py [hunk 15]
```diff
@@ -502,3 +505,112 @@ def gzip_compress_response(raw_content, is_json):
...
+
+def _get_highest_coverage_date(observation_entity_counts_by_date,
+                               facet_ids: Set[str], max_dates_to_check: int,
+                               max_years_to_check: int) -> str | None:
+  """
+  Heuristic for fetching "latest data with highest coverage":
+  Choose the date with the most data coverage from either:
+  (1) last N observation dates
+  (2) M years from the most recent observation date
+  whichever set is greater
```
Comment (chejennifer) [inline server/lib/util.py:570]: nit: whichever set has more dates

File: server/lib/util.py [hunk 16]
```diff
@@ -502,3 +505,112 @@ def gzip_compress_response(raw_content, is_json):
...
+
+  Args:
+    observation_entity_counts_by_date: Part of "dc.get_series_dates" response
+      containing variable observation counts by date and entity
+    facet_ids: (optional) Only consider observation counts from this facet
```
Comment (chejennifer) [inline server/lib/util.py:575]: nit: "from these facets"

File: server/lib/util.py [hunk 17]
```diff
@@ -502,3 +505,112 @@ def gzip_compress_response(raw_content, is_json):
...
+  for observation_entity_counts_by_date in series_dates_response[
+      'datesByVariable']:
+    # Each observation_entity_counts_by_date contains the observation counts by date
+    best_coverage_date = _get_highest_coverage_date(
+        observation_entity_counts_by_date=observation_entity_counts_by_date,
+        facet_ids=facet_ids_set,
+        max_dates_to_check=MAX_DATES_TO_CHECK,
+        max_years_to_check=MAX_YEARS_TO_CHECK)
+    if not best_coverage_date:
+      continue
```
Comment (chejennifer) [inline server/lib/util.py:550]: can you check when regularly fetching observation points if caller expects all requested variables to be in the result?

**Review by chejennifer — APPROVED** [review]
thanks for updating! Just one thing to check before merging and small commenting nits

File: server/lib/util.py (hunk 3 above)
Comment (dwnoble) [reply server/lib/util.py]: Actually, apparently default empty arrays are codacity warning 🤷 

https://pylint.pycqa.org/en/latest/user_guide/messages/warning/dangerous-default-value.html

Reverting to default value `= None`

**Pull request merged**